lib/cmsis_rtos_v1/cmsis_timer.c

```diff
@@ -0,0 +1,126 @@
+/*
+ * Copyright (c) 2018 Intel Corporation
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include <kernel_structs.h>
+#include <cmsis_os.h>
+
+#define ACTIVE 1
+#define NOT_ACTIVE 0
+
+static void zephyr_timer_wrapper(struct k_timer *timer);
+
+struct timer_obj {
+	struct k_timer ztimer;
+	os_timer_type type;
+	u32_t status;
+	void (*callback_function)(void const *argument);
+	void *arg;
+};
+
+K_MEM_SLAB_DEFINE(cmsis_timer_slab, sizeof(struct timer_obj),
+		  CONFIG_MAX_TIMER_COUNT, 4);
+
+static void zephyr_timer_wrapper(struct k_timer *ztimer)
+{
+	struct timer_obj *timer;
+
+	timer = (struct timer_obj *)ztimer;
+
+	(timer->callback_function)(timer->arg);
+}
+
+osTimerId osTimerCreate(const osTimerDef_t *timer_def, os_timer_type type,
+			void *argument)
+{
+	struct timer_obj *timer;
+
+	if (timer_def == NULL) {
+		return NULL;
+	}
+
+	if (type != osTimerOnce && type != osTimerPeriodic) {
+		return NULL;
+	}
+
+	if (k_mem_slab_alloc(&cmsis_timer_slab, (void **)&timer, 100) == 0) {
+		memset(timer, 0, sizeof(struct timer_obj));
+	} else {
+		return NULL;
+	}
+
+	timer->callback_function = timer_def->ptimer;
+	timer->arg = argument;
+	timer->type = type;
+	timer->status = NOT_ACTIVE;
+
+	k_timer_init(&timer->ztimer, zephyr_timer_wrapper, NULL);
+
+	return (osTimerId)timer;
+}
+
+osStatus osTimerStart(osTimerId timer_id, uint32_t millisec)
+{
+	struct timer_obj *timer = (struct timer_obj *) timer_id;
+
+	if (timer == NULL) {
+		return osErrorParameter;
+	}
+
+	if (_is_in_isr()) {
+		return osErrorISR;
+	}
+
+	if (timer->type == osTimerOnce) {
+		k_timer_start(&timer->ztimer, millisec, 0);
+	} else if (timer->type == osTimerPeriodic) {
+		k_timer_start(&timer->ztimer, 0, millisec);
+	}
+
+	timer->status = ACTIVE;
+	return osOK;
+}
+
+osStatus osTimerStop(osTimerId timer_id)
+{
+	struct timer_obj *timer = (struct timer_obj *) timer_id;
+
+	if (timer == NULL) {
+		return osErrorParameter;
+	}
+
+	if (_is_in_isr()) {
+		return osErrorISR;
+	}
+
+	if (timer->status == NOT_ACTIVE) {
+		return osErrorResource;
+	}
+
+	timer->status = NOT_ACTIVE;
+	k_timer_stop(&timer->ztimer);
+	return osOK;
+}
+
+osStatus osTimerDelete(osTimerId timer_id)
+{
+	struct timer_obj *timer = (struct timer_obj *) timer_id;
+
+	if (timer == NULL) {
+		return osErrorParameter;
+	}
+
+	if (_is_in_isr()) {
+		return osErrorISR;
+	}
+
+	if (timer->status == ACTIVE) {
+		timer->status = NOT_ACTIVE;
+		k_timer_stop(&timer->ztimer);
```

Comment:
Do you really need that status field?  I'm all but certain that the k_timer API already has the behavior you're reimplementing here.  Stopping an inactive timer is a silent noop already.

Reply:
Thanks for pointing out that the k_timer API already has this behavior. However, calling _abort_timeout needs me to include quite a few other header files which are otherwise unnecessary. Also, per CMSIS documentation, stopping an inactive timer is an error. Hence, i prefer keeping the status field.